python/lsst/pipe/base/connections.py

```diff
@@ -607,3 +629,111 @@ def iterConnections(connections: PipelineTaskConnections,
+        """
+        # Translate self's DatasetType-keyed, Quantum-oriented mappings into
+        # connection-keyed, PipelineTask-oriented mappings.
+        inputs_by_connection: typing.Dict[str, typing.Tuple[BaseInput, typing.Tuple[DatasetRef, ...]]] = {}
+        outputs_by_connection: typing.Dict[str, typing.Tuple[Output, typing.Tuple[DatasetRef, ...]]] = {}
+        for name in itertools.chain(connections.inputs, connections.prerequisiteInputs):
+            connection = getattr(connections, name)
+            dataset_type_name = connection.name
+            inputs_by_connection[name] = (
+                connection,
+                tuple(self.inputs.get(dataset_type_name, ()))
+            )
+        for name in itertools.chain(connections.outputs):
+            connection = getattr(connections, name)
+            outputs_by_connection[name] = (
+                connection,
+                tuple(self.outputs.get(dataset_type_name, ()))
+            )
+        # Actually call adjustQuantum.
+        adjusted_inputs_by_connection, adjusted_outputs_by_connection = connections.adjustQuantum(
+            inputs_by_connection,
+            outputs_by_connection,
+            label,
+            data_id,
+        )
+        # Translate adjustments to DatasetType-keyed, Quantum-oriented form,
+        # installing new mappings in self if necessary.
+        adjusted_inputs: typing.Optional[NamedKeyDict[DatasetType, typing.List[DatasetRef]]] = None
+        adjusted_outputs: typing.Optional[NamedKeyDict[DatasetType, typing.List[DatasetRef]]] = None
+        for name, (connection, updated_refs) in adjusted_inputs_by_connection.items():
+            dataset_type_name = connection.name
+            if not set(updated_refs).issubset(self.inputs[dataset_type_name]):
+                raise RuntimeError(
+                    f"adjustQuantum implementation for task with label {label} returned {name} "
+                    f"({dataset_type_name}) input datasets that are not a subset of those "
+                    f"it was given for data ID {data_id}."
+                )
+            if adjusted_inputs is None:
+                adjusted_inputs = NamedKeyDict(self.inputs)
+            adjusted_inputs[dataset_type_name] = list(updated_refs)
+        for name, (connection, updated_refs) in adjusted_outputs_by_connection.items():
+            if not set(updated_refs).issubset(self.outputs[dataset_type_name]):
+                raise RuntimeError(
+                    f"adjustQuantum implementation for task with label {label} returned {name} "
+                    f"({dataset_type_name}) output datasets that are not a subset of those "
+                    f"it was given for data ID {data_id}."
+                )
+            if adjusted_outputs is None:
+                adjusted_outputs = NamedKeyDict(self.outputs)
+            adjusted_outputs[dataset_type_name] = list(updated_refs)
+        if adjusted_inputs is not None:
```

Comment:
I cant see how this would be none except and exception

Reply:
It can be `None` of there are no adjustments and hence nothing to iterate over in the earlier loops.

Reply:
See above